src/script/api/script_priorityqueue.cpp

```diff
@@ -0,0 +1,111 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file script_priorityqueue.cpp Implementation of ScriptPriorityQueue. */
+
+#include "../../stdafx.h"
+#include "script_priorityqueue.hpp"
+#include "script_error.hpp"
+#include "../squirrel_helper.hpp"
+#include "../script_instance.hpp"
+#include "../../debug.h"
+
+#include "../../safeguards.h"
+
+
+static bool operator==(const HSQOBJECT &lhs, const HSQOBJECT &rhs)
+{
+	return lhs._type == rhs._type && lhs._unVal.raw == rhs._unVal.raw;
+}
+
+
+ScriptPriorityQueue::~ScriptPriorityQueue()
+{
+	/* Release reference to stored objects. */
+	auto inst = ScriptObject::GetActiveInstance();
+	if (!inst->InShutdown()) {
+		for (auto &i : this->items) inst->ReleaseSQObject(const_cast<HSQOBJECT *>(&i));
+	}
+
+	this->items.clear();
+}
+
+SQInteger ScriptPriorityQueue::Insert(HSQUIRRELVM vm)
+{
+	HSQOBJECT item;
+	int64 priority;
+	sq_resetobject(&item);
+	sq_getstackobj(vm, 2, &item);
+	sq_getinteger(vm, 3, &priority);
+
+	auto inserted = this->items.insert(item);
+	if (inserted.second) {
+		sq_addref(vm, &item); // Keep object alive.
+		this->queue.emplace(priority, item);
+	}
+
+	return SQConvert::Return(vm, inserted.second);
+}
+
+SQInteger ScriptPriorityQueue::Pop(HSQUIRRELVM vm)
+{
+	if (this->IsEmpty()) {
+		ScriptObject::SetLastError(ScriptError::ERR_PRECONDITION_FAILED);
+		sq_pushnull(vm);
+		return 1;
+	}
+
+	HSQOBJECT item = this->queue.top().second;
+	this->queue.pop();
+	this->items.erase(item);
+
+	/* Store the object on the Squirrel stack before releasing it to make sure the ref count can't drop to zero. */
+	auto ret = SQConvert::Return(vm, item);
+	sq_release(vm, &item);
+	return ret;
+}
+
+SQInteger ScriptPriorityQueue::Peek(HSQUIRRELVM vm)
+{
+	if (this->IsEmpty()) {
+		ScriptObject::SetLastError(ScriptError::ERR_PRECONDITION_FAILED);
+		sq_pushnull(vm);
+		return 1;
+	}
+
+	return SQConvert::Return(vm, this->queue.top().second);
+}
+
+SQInteger ScriptPriorityQueue::Exists(HSQUIRRELVM vm)
+{
+	HSQOBJECT item;
+	sq_resetobject(&item);
+	sq_getstackobj(vm, 2, &item);
+
+	return SQConvert::Return(vm, this->items.find(item) != this->items.end());
+}
```

Comment:
With a quick look I didn't see any pathfinder code actually using the Exists() function as defined in the existing squirrel implementations. I cannot remember why it was added. The FibonacciHeap implementation already documents this as O(n), which is bad. I'd suggest to leave this function out which means the class doesn't need to keep the items set.

Reply:
Indeed, at least the road pathfinder lib doesn't use Exists. I added it because the AI libs have it, but I can of course easily scrap it if you don't know any use for it.

Reply:
I'd scrap this for now and only readd if there is a compelling reason